apps/build-onchain-apps/contracts/src/SignatureMintERC721.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+import {NFTUtilities} from "./NFTUtilities.sol";
+
+// Errors
+error InsufficientFunds();
+
+/**
+ * @title SignatureMintERC721
+ * @dev contract that allows a user to mint a ERC721 for free with a cryptographically signed message.
+ *      This is useful for mints where you want to allow users to mint for free based on a signature from
+ *      your backend API.  This is an alternative approach to a merkel tree which is fully on-chain.
+ */
+contract SignatureMintERC721 is ERC721 {
+    using ECDSA for bytes32;
+    using MessageHashUtils for bytes32;
+
+    // Contract Params
+    uint256 public paidMintCost = 0.0001 ether;
+    address public freeMintSigner; // Wallet Address for the authorized wallet to distribute free mints.
+    uint256 public currentId = 0; // Current NFT id
+    mapping(address => uint256) public usedFreeMints; // Mapping to store free mints
+
+    // Setup contract Metadata
+    string public imageUri = "ipfs://QmQRzw5MtNQcdjg9rZ8MYugAz6WjaBdvsuNShKTLYNJdj7/coffee_dog.png";
+    NFTUtilities.CollectionMetadata public metadata =
+        NFTUtilities.CollectionMetadata("CoffeeDog", "Buy a dog a coffee", "CDOG", imageUri, "");
+
+    // Events
+    event Mint(address indexed buyer, uint256 tokenId, uint256 cost);
+
+    constructor(address signer) ERC721(metadata.name, metadata.symbol) {
+        freeMintSigner = signer;
+    }
+
+    /**
+     * @dev Function to mint for free using a wallet signature
+     * @param  account that will receive the NFT
+     * @param  signature that authorizes this transaction to mint for free.
+     */
+    function freeMint(address account, bytes memory signature) external {
+        require(this.verifySignature(account, signature) == true, "Unable to verify Signature");
+
+        // Check if the free mint has already been used (1 per wallet)
```

Comment:
might want to make this configurable?

Reply:
again, this is meant to be an example users can configure for themself.  I didn't want to make the constructor super busy for this example.  @Zizzamia  any opinions here if we want to make the contracts constructor super configurable?

Reply:
for example, mint price, should that be configurable via constructor or just hardcoded in the contract.  Back to our earlier point, we aren't trying to compete with OpenZepplin here so I don't think we need to fully future-proof our contracts.

Reply:
Right right, and I think it goes back to also the point on "how long does it take to understand and update a contract". 

Mint price in this case is 0, would be the case this contract will be used on a case is not 0? And if that's the case, a configuration will make an easier upgrade path vs hardcoded.

My guess at least for price, have that configurable seems reasonable.

Reply:
Okay taking in all the points here, I think we should keep it simple for now.  It will be trivial for a developer to make these values configurable.  I think the readability is more important at this phase.  This is an easy one-way door if we want to revisit at a later point.